Feature: Import OTPs from ZIP archive

z64me opened the pull request:
# PR Details

Users can now reimport OTPs previously exported from Raivo. Closes #22

![Screen_Recording_2023-03-31_at_11 22 52_PM](https://user-images.githubusercontent.com/45853441/229264852-c31e6bac-864a-428b-9b07-8d52753cae44.gif)

## Description

For every file related to exporting OTPs, I have created an equivalent for importing OTPs, e.g. I added `MainDataImportView`/`DataImportFeature`/etc to complement the existing `MainDataExportView`/`DataExportFeature`/etc.

The import process spans two steps: the user selects the ZIP file to import, and then they are prompted to enter the password required for decryption.

## Type of change

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# Checklist:

- [x] My code follows the code style of this project.
- [x] My change generates no new warnings
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.

z64me committed:
Feature: Import OTPs from ZIP archive

Users can now reimport OTPs previously exported from Raivo. Closes #22

tijme commented on Raivo/Forms/MiscellaneousForm.swift:
```diff
@@ -287,7 +287,7 @@ class MiscellaneousForm {
             section.tag = "data"
             section.hidden = Condition(booleanLiteral: !authenticated)
-            section.footer = HeaderFooterView(title: "Your data will be exported in an encrypted ZIP archive (using your encryption password).")
+            //section.footer = HeaderFooterView(title: "Your data will be exported in an encrypted ZIP archive (using your encryption password).")
```
Why is this commented?

z64me replied:
That was me experimenting with how that section looks with/without a footer. Each option's title and respective page are self explanatory, but I couldn't make up my mind on whether to delete or rephrase the footer, so I commented it out for now.

tijme reviewed (commented):
Awesome pull request @z64me 🚀 . Users of Raivo will really like it, it's one of the most requested features!

I have some feedback points. The main one being that I don't think the old passwords should be deleted. I think users should be instructed to delete those themselves, to prevent accidental deletes.

z64me commented on Raivo/Features/DataImportFeature.swift:
```diff
@@ -0,0 +1,111 @@
+            deleteFile(filename)
+            
+            // Clean up old passwords
+            deleteAllPasswords()
```
What are your thoughts on keeping all existing passwords, but overwriting those where the issuer and username match one from the backup? This way, users aren't left with duplicates each time they reimport. I also really like the feature proposed in #212 as a potential solution to this problem.

tijme replied:
I quite like the recycle bin, but it's also quite a lot of work to implement.

Anyhow, personally I quite like your suggestion to 'replace' tokens by username/issuer. However, I'm pretty sure there are less tech-savvy users out there who don't pay close attention and end up losing old tokens because of this. Therefore, I would suggest playing it safe and just warning users to manually delete old tokens first. Then the user cannot do anything wrong by accident.

z64me replied:
That makes sense: losing a token you can't get back is a way bigger problem than having to clean up a few duplicate/old OTPs. :)

All requested changes are now live in the latest commit. I've added an optional parameter called `shouldDeleteOldPasswords` in case we revisit this after implementing a recycle bin.

z64me commented on Raivo/Features/DataImportFeature.swift:
```diff
@@ -0,0 +1,111 @@
+            return ("Import Successful", "New OTPs were successfully imported from the ZIP archive.")
+        } catch {
+            // Return error message
+            return ("Import Failed", "Failed to import OTPs from ZIP archive.")
```
Good idea! I will add this.

z64me committed:
OTP Import: More descriptive error messages

Error messages added and tested for specific failure conditions:
- Incorrect password
- Missing JSON
- Invalid JSON structure
- Empty JSON array

z64me committed:
OTP import: Optional password deletion + phrasing

tijme commented:
Hmm, I'm testing it personally in TestFlight right now, but unfortunately it states that the password is incorrect. So this feature does not seem to work with all ZIP archives yet. What I did to reproduce;

* Update app
* Export ZIP
* Logout
* Login
* Import ZIP

What's strange is that this *does* work in the simulator, just not on my iPhone.

z64me commented:
Could it be an issue with file paths or permissions? What kind of output do you get from `SSZipArchive.isFilePasswordProtectedAtPath()`?

tijme commented:
Wow, that took a while to debug for me. But you were right. This was indeed a permissions error on the path. Learned something new today, as I was debugging with string paths instead of URL objects.

tijme commented:
Will push a new version to TestFlight later today. Thanks!

tijme commented:
This was just released to the AppStore. Thanks again.

z64me commented:
I downloaded the latest update earlier and it works great. 👍 Thank you for maintaining Raivo. It is great software. ❤️

dkorunic commented:
I have been able to reproduce an sync issue with ZIP imports. Exporting everything to a ZIP file, signing out and creating a new master password, I've reimported successfully all TOTP seeds, but on a secondary device synced through iCloud only ~40% of the seeds are visible. I was able to resync manually all of seeds by editing and saving one by one, but I think that there is some kind of race with batch syncing after import. During syncing 54 seeds, I've observed only ~26 being synced in the sync status progress.

reterkha commented:
> since the developers leave no way to contact them, have closed the issue section here on github I want to raise the ongoing issue that raivo otp has without my consent deleted all entries in my application. This is a critical issue, locking out businesses out of their services. 
> 
> 
> 
> How can I restore my data?

You can't we've all moved on.  I didn't have any backups and lost everything.   

This app is dead now.  The new devs have lost their investment.
